common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts

```diff
@@ -1,15 +1,41 @@
 import * as xml2js from 'xml2js';
-import { LDMLKeyboardXMLSourceFile } from './ldml-keyboard-xml.js';
+import { LDMLKeyboardXMLSourceFile, LKImport } from './ldml-keyboard-xml.js';
 import Ajv from 'ajv';
 import { boxXmlArray } from '../util/util.js';
+import { CompilerCallbacks } from 'src/util/compiler.js';
+import { constants } from '@keymanapp/ldml-keyboard-constants';
 
 export default class LDMLKeyboardXMLSourceFileReader {
+  callbacks: CompilerCallbacks;
+
+  constructor(callbacks : CompilerCallbacks) {
+    this.callbacks = callbacks;
+  }
+
+  readImportFile(version: string, subpath: string): Buffer {
+    // TODO-LDML: sanitize input string
+    let importPath = new URL(`../import/${version}/${subpath}`, import.meta.url);
+    // TODO-LDML: support baseFileName?
+    return this.callbacks.loadFile(importPath.pathname, importPath);
+  }
+
   /**
    * xml2js will not place single-entry objects into arrays.
    * Easiest way to fix this is to box them ourselves as needed
    * @param source any
    */
   private boxArrays(source: any) {
+    if (source?.keyboard) {
+      if (!source.keyboard.keys) {
+        source.keyboard.keys = {
+          key: [],
+          flicks: [],
+        };
+      }
+      if (!source.keyboard.keys?.import) {
+        source.keyboard.keys.import = [];
```

Comment:
This will crash if `source.keyboard.keys` is nullish

Reply:
If it's here it's still an issue. Will 👀 look

Reply:
Really? doesn't seem to:

```shell
$ node
> source = {keyboard: {}};
{ keyboard: {} }
>     if (source?.keyboard) {
...        if (!source.keyboard.keys) {
...          source.keyboard.keys = {
...            key: [],
...            flicks: [],
...          };
...        }
...        if (!source.keyboard.keys?.import) {
...          source.keyboard.keys.import = [];
...  
... }
... }
[]
> source
{ keyboard: { keys: { key: [], flicks: [], import: [] } } }
```

line 30 (just prior)  sets source.keyboard.keys if it was otherwise nullish. there's an extra questionmark though.

```suggestion
      if (!source.keyboard.keys.import) {
        source.keyboard.keys.import = [];
```